core/opendaq/modulemanager/src/module_manager_impl.cpp

```diff
@@ -687,53 +697,57 @@ void ModuleManagerImpl::attachStreamingsToDevice(const MirroredDeviceConfigPtr&
-
-            if (protocolPriority > primaryProtocolPriority)
-                primaryProtocolParams = {cap.getPrefix(), protocolPriority};
+            if (const auto detachedSourceIt = detachedStreamingSources.find(primaryConnectionString);
+                detachedSourceIt != detachedStreamingSources.end())
+            {
+                streaming = detachedSourceIt->second;
+            }
         }
-        else
+
+        // connect if source is not found in detached sources
+        if (!streaming.assigned())
         {
-            primaryProtocolParams = {cap.getPrefix(), protocolPriority};
+            streaming = createStreaming(primaryConnectionString, cap);
+            if (!streaming.assigned())
+                continue;
         }
 
-        LOG_I("Device {} added streaming connection {}", device.getGlobalId(), streaming.getConnectionString());
+        prioritizedStreamingSourcesMap.insert_or_assign(protocolPriority, streaming);
+    }
+
+    // add streaming sources ordered by protocol priority
+    for (const auto& [_, streaming] : prioritizedStreamingSourcesMap)
+    {
         device.addStreamingSource(streaming);
+        LOG_I("Device {} added streaming connection {}", device.getGlobalId(), streaming.getConnectionString());
 
         streaming.addSignals(signals);
         streaming.setActive(true);
     }
 
     // set the streaming source with the highest priority as active for device signals
-    if (primaryProtocolParams.has_value())
+    if (!prioritizedStreamingSourcesMap.empty())
     {
-        auto primaryProtocolPrefix = primaryProtocolParams.value().first.toStdString();
-        for (const auto& streaming : device.getStreamingSources())
+        auto streaming = prioritizedStreamingSourcesMap.begin()->second;
+        auto connectionString = streaming.getConnectionString();
+        for (const auto& signal : signals)
         {
-            auto connectionString = streaming.getConnectionString();
-            if (connectionString.toStdString().find(primaryProtocolPrefix) == 0)
-            {
-                for (const auto& signal : signals)
-                {
-                    if (!signal.getPublic())
-                        continue;
-                    MirroredSignalConfigPtr mirroredSignalConfigPtr = signal.template asPtr<IMirroredSignalConfig>();
-                    if (!mirroredSignalConfigPtr.getActiveStreamingSource().assigned() || overrideActiveSourceForSignal)
-                        mirroredSignalConfigPtr.setActiveStreamingSource(connectionString);
-                }
-            }
+            if (!signal.getPublic())
+                continue;
+            MirroredSignalConfigPtr mirroredSignalConfigPtr = signal.template asPtr<IMirroredSignalConfig>();
+            if (!mirroredSignalConfigPtr.getActiveStreamingSource().assigned() || overrideActiveSourceForSignal)
+                mirroredSignalConfigPtr.setActiveStreamingSource(connectionString);
```

Comment:
Does this mean another source might be active before the highest priority one? It might make more sense to first connect the highest priority source in general. If it's bundled with the configuration connection, I'd assume it's anyhow the highest priority one.

As for native, it should also be possible to connect config-only, without streaming. And if that's possible, I'd assume that all streamings are "detached" anyhow.

Reply:
> Does this mean another source might be active before the highest priority one? 

When the `ModuleManagerImpl::attachStreamingsToDevice` is called for some device and either the "MIN_HOPS" or "FALLBACKS"  heuristic is selected, certain signals may already have an active streaming source assigned  - the highest priority source of the nested device. If the "FALLBACKS" heuristic is selected, this assignment gets overwritten by the highest priority source of the parent device. However, it doesn't really make sense to provide the "FALLBACKS" heuristic until we define and implement a procedure for automatically reconfiguring streaming in cases where the previously selected active streaming source disconnects or fails. Therefore, for now, I would limit the possible heuristics to these three: "MIN_HOPS," "MIN_CONNECTIONS," and "DISCONNECTED."

> As for native, it should also be possible to connect config-only, without streaming. And if that's possible, I'd assume that all streamings are "detached" anyhow.

Regarding the native streaming for native config devices - it's automatically connected by the client module during the config connection step to utilize the same pipe for both streaming and configuration. However, to ensure proper handling of streaming properties in the module manager, previously established streaming objects associated with the device are detached (removed) and re-added only if the corresponding protocol is enabled by the streaming properties. To streamline the implementation, we could remove the automatic creation of native streaming from the module implementation. Rather than being automatically created along with the config connection, the native streaming object for native config devices would be instantiated by the module manager, just like any other type of streaming. It would then operate in a separate pipe. This approach removes the necessity for custom handling of this source in the module manager, such as detaching it from the device and re-adding it.